### PR TITLE
Use regular expression instead of split when replacing authorized_keys

### PR DESCRIPTION
**Pull request by Foxlik**
### What does this PR do?

In some cases (e.g. Ubuntu cloud-init) authorized_keys may contain quoted spaces as part of options:
```
no-port-forwarding,no-agent-forwarding,no-X11-forwarding,com‌​mand="echo 'Please login as the user \"ubuntu\" rather than the user \"root\".';echo;sleep 10" ssh-rsa xxxxxx
```
which confused the split in _replace_auth_key() in modules/ssh.py.

I kind of feel, that there may be a much nicer solution to this, just can't come up with it, so please feel free to reiterate. On the other hand, it simplifies the detection a little. 

Oh and please, doublecheck my regular expression ... it works for all the cases I could think of, but ... 

### What issues does this PR fix or reference?

#39854

### Previous Behavior

Salt would report there is a change in the state on every run, but would change nothing. 

### New Behavior

Now Salt correctly replaces the line.

### Tests written?

No 

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

File: salt/modules/ssh.py [hunk 1]
```diff
@@ -149,15 +149,8 @@ def _replace_auth_key(
...
-                comps = line.split()
-                if len(comps) < 2:
-                    # Not a valid line
-                    lines.append(line)
-                    continue
-                key_ind = 1
-                if comps[0][:4:] not in ['ssh-', 'ecds']:
-                    key_ind = 2
-                if comps[key_ind] == key:
+                comps = re.findall(r'((.*)\s)?(ssh-[a-z0-9-]+|ecdsa-[a-z0-9-]+)\s([a-zA-Z0-9+/]+={0,2})(\s(.*))?', line)
```
Comment (terminalmage) [inline salt/modules/ssh.py]: The first two character classes in this regex contains a `-` at the end. This would match on `ssh--rsa`, or `ecdsa--sha2-nistp256` which I don't think is intended.

Comment (Foxlik) [inline salt/modules/ssh.py]: I'd say that compared to previous solution it's a little more precise. Thoug, I think the 'unique line id', if I may call it so, is the key itself.
I want salt (though I may not be average salt user) to see, if there is such a key in the file, and if so, make sure it has the options, type and comment I asked it to have. 
From my point of view, any key with double dash or ecdsa---foo---bar---baz, or pretty much anything else gets fixed this way.
This part of the expression really serves rather as a delimiter to figure out which part of the line really is the key. Though truth be told, thinking about it now, this could probably be done better, as having ssh-rsa somewhere in the options, would confuse the regexp as well.

Comment (terminalmage) [inline salt/modules/ssh.py]: Could you elaborate a little more on the logic behind this regex? It's quite complex and I don't grok what it is you're doing here, to be honest.

Comment (Foxlik) [inline salt/modules/ssh.py]: I really tried to follow the idea of the original code. Which just split the line by spaces and tried to match 1st part for (ssh-|ecds). If it matched, then treat the line as if second part is the key, if it did not match, use third part as the key. 

`((.*)\s)?` the line may start with options. it may contain pretty much anything, including quotes, escaped quotes, and so on, followed by a space.

`(ssh-[a-z0-9-]+|ecdsa-[a-z0-9-]+)\s` next is the key type, which really the only known part of the line. Presently it has to start with ssh- or ecds-. Followed by a space.

`([a-zA-Z0-9+/]+={0,2})` now we have to match the key. It is base64 encoded.

`(\s(.*))?` the last is a comment. Which again is pretty much free-form. 

So in short, now I'm looking for `(ssh-|ecdsa-)` as i know that is followed by a space and a base64 encoded key.

To compare, this is how openssh itself does it: https://github.com/openssh/openssh-portable/blob/52763dd3fe0a4678dafdf7aeb32286e514130afc/auth2-pubkey.c#L774 ... I really don't feel like porting that to python and not causing any damage myself.

Comment (cachedout) [inline salt/modules/ssh.py]: @terminalmage Can you please follow-up here when you get a chance?

Comment (terminalmage) [inline salt/modules/ssh.py]: OK, thanks for the explanation. Given the complexity of the regular expression, I would feel a lot better if we added a unit test for this.

Comment (cachedout) [inline salt/modules/ssh.py]: @Foxlik Do you feel like you could make a unit test for this?

Comment (Foxlik) [inline salt/modules/ssh.py]: @cachedout to be honest, I don't feel qualified for that. I'm not much of a developer and as this function is in the business of reading and writing a file, I'd probably just mess things up. The same goes for refactoring the matching and reading/writing.
Though if you could give me a pointer where to steal the test from, I could give it a shot. 
Or I can come up with some test data if you (or someone else) would be willing to integrate it.

**Comment by cachedout** [comment]
@rallytime Is this something you might be able to put together a quick automated test for?

**Comment by rallytime** [comment]
@cachedout Sure. I'll try to get to that this afternoon or tomorrow morning.

**Comment by rallytime** [comment]
Hi @Foxlik - I have added a unit test for this change in #40379. Please take a look and let me know if there are any use cases I have missed. I tried to get everything you mentioned in your PR and related issue.
